src/main/java/com/blog/stream/pagination/PageSpliterator.java

```diff
@@ -1,108 +1,189 @@
             final int pageSize,
-            final PageFetcher<T> pageFetcher) {
-        this.pageNumber = pageNumber;
+            final BiFunction<Integer, Integer, P> pageFetcher,
+            final Function<P, List<T>> itemExtractor,
+            final Function<P, Integer> totalPagesExtractor
+            ) {
+        this.pageNumber = new AtomicInteger(pageNumber);
         this.count = count;
         this.pageSize = pageSize;
         this.pageFetcher = pageFetcher;
+        this.itemExtractor = itemExtractor;
+        this.totalPagesExtractor = totalPagesExtractor;
     }
 
-    static <R> PageSpliterator<R> create(final int count, final int pageSize, final PageFetcher<R> pageFetcher) {
-        return new PageSpliterator<>(0, count, pageSize, pageFetcher);
+    /**
+     * construct a spliterator for sequential, or parallel if a good estimate of the count is known ahead of time
+     * @param count total pages expected
+     * @param pageSize size of page
+     * @param pageFetcher function given page number returns a page object of type P
+     * @param itemExtractor function given a page object of type P return a list of type T
+     * @param <P> page object type
+     * @param <T> item type
+     * @return created spliterator
+     */
+    public static <P, T> PageSpliterator<P, T> create(final int count, final int pageSize, final BiFunction<Integer, Integer, P> pageFetcher, Function<P, List<T>> itemExtractor) {
+        return PageSpliterator.create(0, count, pageSize, pageFetcher, itemExtractor);
     }
 
-    @Override
-    public boolean tryAdvance(final Consumer<? super T> action) {
-        Page<T> page = pageFetcher.fetch(pageNumber, pageSize);
-        page.forEach(action);
-
-        //in parallel mode this section will always be run on the last page
-        pageNumber++;
-        return !page.isLast();
+    public static <P, T> PageSpliterator<P, T> create(final int pageNumber, final int count, final int pageSize, final BiFunction<Integer, Integer, P> pageFetcher, Function<P, List<T>> itemExtractor) {
+        return new PageSpliterator<>(pageNumber, count, pageSize, pageFetcher, itemExtractor, null);
     }
 
     /**
-     * Attempt to split the spliterator. This is easy as by definition the result set is defined as paged buckets
-     *
-     * If the stream is run in parallel the spliterator will be split as below
-     * (4 queries for a given pageSize/count, top row is page number)
-     * |   0    |   1    |   2    |   3    |
-     * | Parent |        |        |        |  No Splits
-     * | Child  | Parent |        |        |  First Split
-     * | Child  | Child  | Parent |        |  Second Split
-     * | Child  | Child  | Child  | Parent |  Third Split
-     *                                        No more splits (return null)
-     *
-     *  If run sequentially, try split will not be called and the tryAdvance handles paging
-     * @return Child spliterator whose role is to iterate over one set and die.
```

Comment:
Probably worth leaving this javadoc as is, the null aspect of the class can be confusing

Reply:
ooo. sorry, good catch